reccmp/isledecomp/compare/core.py

```diff
@@ -335,29 +325,50 @@ def _add_match_in_array(
 
             data_type = self.cv.types.keys[node.data_type.key.lower()]
 
-            if data_type["type"] == "LF_ARRAY":
-                array_element_type = self.cv.types.get(data_type["array_type"])
-
-                assert node.data_type.members is not None
-
-                for array_element in node.data_type.members:
-                    orig_element_base_addr = match.orig_addr + array_element.offset
-                    recomp_element_base_addr = match.recomp_addr + array_element.offset
-                    if array_element_type.members is None:
-                        _add_match_in_array(
-                            f"{match.name}{array_element.name}",
-                            array_element_type.key,
-                            orig_element_base_addr,
-                            recomp_element_base_addr,
-                        )
-                    else:
-                        for member in array_element_type.members:
-                            _add_match_in_array(
-                                f"{match.name}{array_element.name}.{member.name}",
-                                array_element_type.key,
-                                orig_element_base_addr + member.offset,
-                                recomp_element_base_addr + member.offset,
-                            )
+            if data_type["type"] != "LF_ARRAY":
+                continue
+
+            # Check whether another orig variable appears before the end of the array in recomp.
+            # If this happens we can still add all the recomp offsets, but do not attach the orig address
+            # where it would extend into the next variable.
+            upper_bound = match.orig_addr + match.size
+            if (next_orig := self._db.get_next_orig_addr(match.orig_addr)) is not None:
+                if next_orig < upper_bound:
+                    logger.warning(
+                        "Array variable %s at 0x%x is larger in recomp",
+                        match.name,
+                        match.orig_addr,
+                    )
+                    upper_bound = next_orig
+
+            array_element_type = self.cv.types.get(data_type["array_type"])
+
+            assert node.data_type.members is not None
+
+            for array_element in node.data_type.members:
+                orig_element_base_addr = match.orig_addr + array_element.offset
+                recomp_element_base_addr = match.recomp_addr + array_element.offset
+
+                if array_element_type.members is None:
+                    # If array of scalars
+                    name = f"{match.name}{array_element.name}"
+                    orig_addr = orig_element_base_addr
+                    recomp_addr = recomp_element_base_addr
+
+                    dataset[recomp_addr] = {"name": name}
+                    if orig_addr < upper_bound:
+                        orig_by_recomp[recomp_addr] = orig_addr
```

Comment:
How about adding back the helper function and passing `upper_bound` as an additional parameter? I like the added comments what each case corresponds to.